### DISCUSSION TITLE
Localization platform for TerriaJS

### DISCUSSION THREAD

**Discussion by arky** [TerriaJS/terriajs#5072] [Q&A] [answered]
Does TerriaJS use a localization platform such as Transfix or Crowdin ? We would like contribute localization strings for Myanmar, Khmer, Vietnamese, Thai and Lao languages. Please kindly advise. Thanks 

https://github.com/TerriaJS/terriajs/issues/3785

**Comment by KeyboardSounds** [comment, marked as answer]
Hi @arky,

We don't use a localization platform as such, but we do support localization, and we would absolutely love to have translations for as many languages as we can. Our localization strings are specified like this: [v7](https://github.com/TerriaJS/terriajs/blob/master/lib/Language/en/translation.json) and [v8](https://github.com/TerriaJS/terriajs/blob/next/lib/Language/en/translation.json), depending on which version of Terria you're interested in.

The best way to contribute a translation would be to copy the file for English and then translate each of the strings in there. If you've got any more questions or you'd like some help, please let me know :)

**Reply by MoRadwan74** [reply]
Hi @KeyboardSounds,

I made a [PR](https://github.com/TerriaJS/terriajs/pull/5241) about adding the Arabic language to the app with a toggle-language dropdown added to the menu buttons alongside RTL styles handling. Also, provide a guide about how to add a new language. Would you review it, please?

**Reply by arky** [reply]
>Also, provide a guide about how to add a new language. Would you review it, please?

You can add a new language on translation platform using the 'Start New Language' button 
https://hosted.weblate.org/new-lang/terriajs/terriajs/
![image](https://user-images.githubusercontent.com/808895/107929016-a39d0700-6fab-11eb-8721-78956ec11b2e.png)

**Reply by MoRadwan74** [reply]
@arky That's great! It will make it easy to add any language for sure. I have added some RTL styles to the PR for RTL languages like Arabic and Hebrew.

**Reply by arky** [reply]
@MoRadwan74 I have added [Arabic](https://hosted.weblate.org/projects/terriajs/terriajs/ar/) if you would like to translate.

**Reply by MoRadwan74** [reply]
@arky Yes, sure! Thanks.

**Comment by arky** [comment]
We have already translated a bunch of strings for SE. Asian languages. At the moment, we have hit the limit on number of strings we can host on Transfix. We would love to push these translated strings up stream so we can streamline these translations.

@prustar

**Comment by zoran995** [comment]
@KeyboardSounds this is something that definitely should be considered, with just two languages it becomes difficult to keep them up to date and entirely in sync (speaking from experience on our project) and terriajs can expect much more. 
Crowdin and Transifex have a free plan for opensource projects.
@arky I believe you meant Transifex instead of Transfix

**Reply by KeyboardSounds** [reply]
@zoran995 it would be great to hear more about the specific problems you're having with translating Terria. We probably aren't going to use an external i18n platform at this stage, but we might be able to do other things to make it easier.

**Reply by zoran995** [reply]
We started with manual translation since I extracted the strings in the master branch. When having a team for translation that is not really in development, it easily gets corrupt, so I have to be really careful with a file when dropping it back in. 
Also, when something in terriajs changes it ofter requires updating the translation files and it is not easy to always track the changes and update all translation files, either add new strings or update existing. The platform would make it a lot easier since it can be configured to make the translation unapproved so somebody check it. A few days ago I got Cyrillic and Latin version out of sync, which requires just conversion of characters (just realised that I should write a function for that and include it in the build process).

**Comment by arky** [comment]
Yes, @zoran995 I meant Transifex. 

We are using Transifex for our work. They do provide a generous free tier for open source / non-profits but when you are working with multiple projects/ languages (Khmer, Thai, Lao, Vietnamese and Myanmar). You will quickly hit the string limit. 

If TerriaJS a platform, then we can contribute translations to it directly as we have dedicated staff to do the localization for these languages.

**Comment by meh9** [comment]
@arky @zoran995 
Thank you for the interest in discussing and contributing translations. 

Unfortunately, even though TerriaJS is open source, we don't qualify for the free open source tiers on either of the translation platforms above since we have commercial contracts for delivering TerriaJS applications. 

At the moment we are also not able to prioritise the development work required to use the platforms - however we may re-evaluate this in the future.

In the meantime if you are able and willing we would love to accept PR's with translations based on our existing methods described above.

Thanks!

**Reply by arky** [reply]
> @arky @zoran995
> At the moment we are also not able to prioritise the development work required to use the platforms - however we may re-evaluate this in the future.
> 
> In the meantime if you are able and willing we would love to accept PR's with translations based on our existing methods described above.

Am going to investigate a possible workflow to enable our staff to help with localization. Would be happy to send the PR's in next few weeks. 

Could you tell me which stable and development branches/tags we should target for localization. And also advice cadence of pushing PR's for localization files. Thanks.

**Reply by zoran995** [reply]
@arky, I would target a `next` branch known as the v8 version, TerriaJS team is working on porting all their implementations to the v8, and it will completely replace the `master`. Someone from the team can comment on the expected date, but I expect v8 to become officially stable in few months.
V8 also has much better internationalization support (i.e., changing strings without recompiling the code), but keep in mind that many source strings might differ between `master` and `next` branch.

**Reply by arky** [reply]
@zoran995 Thanks, I am hoping we'll get started on TerriaJS translating on [weblate](https://hosted.weblate.org/projects/terriajs-l10n/terriajs/) once it goes through the approval process.

**Reply by arky** [reply]
> At the moment we are also not able to prioritise the development work required to use the platforms - however we may re-evaluate this in the future.
> 
@meh9  I believe you don't need to make any development to integrate a localization system into TerriaJS.  I have taken time to setup, test, add screenshots for translators, translating strings on [TerriaJS l10n project on weblate](https://hosted.weblate.org/projects/terriajs-l10n/terriajs/), you'll only administer and manage the project  as described in the bug #5126

**Comment by prustar** [comment]
@meh9 -- the ODM is a not-for-profit and we are using all open source application. TerriaJS has been fundamental to our implementation of our database. Our target audience are non-English speaking and it would be invaluable to have a systemic approach to managing ongoing translations into the future. 

Wondering how we could help support you in making this happen as Asia has a large user base and it would be good to increase the use of our platform as well as promote the use of TerriaJS more broadly.

**Comment by arky** [comment]
TerriaJS project is eligible for libre hosting from Weblate.org. I have requested the project to be added to weblate and also done necessary testing. Our contributors would be able to translate using online platform more efficiently than regularly. 

https://hosted.weblate.org/projects/terriajs-l10n/

In order to complete the setup, someone from the TerriaJS admin team needs to the join the project and complete the last step in Weblate.org community checklist. https://hosted.weblate.org/guide/terriajs-l10n/terriajs/. 

https://github.com/TerriaJS/terriajs/issues/5126

**Comment by arky** [comment]
Wonder if anyone from @terriaJS developer team could join and manage TerriaJS project on weblate ? It would help the community contribute localization more effectively. Thanks!

https://hosted.weblate.org/projects/terriajs-l10n/terriajs/

https://github.com/TerriaJS/terriajs/issues/5126

**Comment by arky** [comment]
Thank you everyone, TerriaJS official adopted weblate as their translation platform. Please join and translate TerriaJS into your language. 

https://github.com/TerriaJS/terriajs#big-thanks

**Comment by arky** [comment]
To maximize the efficiency of translation, I have marked all the strings that are commonly presented to the users in TerriaJS Menus,  workbench, Story editor with higher priority. These strings would be present to the translators first so they can be quickly translated. https://hosted.weblate.org/translate/terriajs/terriajs/en/?q=label:%22UI%22

**Reply by imakihi** [reply]
@arky Thank you for your great work. I just started Japanese translation. I am not a web developer and would like to know how our app developer can localize Terria.js Next branch with my translation. Can you explain how this work??
Thank you!
Hiroo
